pkg/util/steps/runner.go

```diff
@@ -0,0 +1,177 @@
+package steps
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+// StepResultHandler is the interface for the handler of Runner.
+//
+// OnStepError is called when a step or condition fails and returns the error to
+// fail with. This error may be different than the error passed in originally.
+//
+// ShouldRetryStep is called when an Action fails to check whether it should be
+// retried. A return value of (true, nil) indicates it should be reran, all
+// other return values will cause the Action to fail. If the check itself causes
+// an error, it may return that instead, but it should also log the original
+// error. ShouldRetryStep will be ran in the same timeout deadline as the step
+// that failed.
+type StepResultHandler interface {
+	OnStepError(ctx context.Context, err error, step Step) error
+	ShouldRetryStep(ctx context.Context, err error, step ActionFunction) (bool, error)
+}
```

Comment:
I think I'm -1 on the existence of this interface.

I think OnStepError can be removed as follows:
1) when an action or condition fails, have the Runner always log an error with the step name and error (this changes the current logging behaviour but I think that's OK).
2) when the Runner returns with an error, have the caller call gatherFailureLogs.

I think ShouldRetryStep can be removed as follows: define a new adapter type which implements Action which separates out the retry functionality.  That adapter can be applied to the Actions that actually need the retry.  Those are probably just deployStorageTemplate, attachNSGsAndPatch and deployResourceTemplate.

Reply:
Done!